File: resources/sequence.py
import seq as sq

class Sequence:
	def __init__(self, string):
		self.setSequence(sq.createSeq(string))
		self.setLimits()
		
	def setLimits(self):
		self.lower = list(self.sequence)[0]
		self.upper = list(self.sequence)[-1]

	def setSequence(self, seq):
		self.sequence = seq

	def getSequence(self):
		return self.sequence

	def getString(self):
		return sq.toString(self.sequence, self.lower, self.upper)

	def conSequence(self, g): #Convolucion
		res = sq.conSeq(self.sequence, self.lower, self.upper, g, g.lower, g.upper)
		return Sequence(sq.toString(res, list(res)[0], list(res)[-1]))

	def __add__(self, other): #f(n) + g(n) ó f(n + C)
		if (type(other) == type(int())): #Cuando es desplazamiento
			res = sq.desSeq(self.sequence, self.lower, self.upper, other)
			return Sequence(sq.toString(res, list(res)[0], list(res)[-1]))
		l = min(self.lower, other.lower)
		u = max(self.upper, other.upper)
		res = sq.addSeqs(self.sequence, other.sequence, l, u)
		return Sequence(sq.toString(res[0], res[1], res[2]))

	def __iadd__(self, other): #f(n) += f(n)
		l = min(self.lower, other.lower)
		u = max(self.upper, other.upper)
		res = sq.addSeqs(self.sequence, other.sequence, l, u)
		return Sequence(sq.toString(res[0], res[1], res[2]))

	def __sub__(self, other): #f(n) - g(n) ó f(n - C)
		if (type(other) == type(int())): #Cuando es desplazamiento
			res = sq.desSeq(self.sequence, self.lower, self.upper, -other)
			return Sequence(sq.toString(res, list(res)[0], list(res)[-1]))
		l = min(self.lower, other.lower)
		u = max(self.upper, other.upper)
		res = sq.subSeqs(self.sequence, other.sequence, l, u)
		return Sequence(sq.toString(res[0], res[1], res[2]))

	def __isub__(self, other): #f(n) -= f(n)
		l = min(self.lower, other.lower)
		u = max(self.upper, other.upper)
		res = sq.subSeqs(self.sequence, other.sequence, l, u)
		return Sequence(sq.toString(res[0], res[1], res[2]))
		
	def __mul__(self, other): #f(n) · g(n) ó f(n) · C
		if(type(other) == type(int())): #Cuando es amplificació o atenuación
			return Sequence(sq.toString(sq.ampSeq(self.sequence, self.lower, self.upper, other), self.lower, self.upper))
		else:
			l = min(self.lower, other.lower)
			u = max(self.upper, other.upper)
			res = sq.mulSeqs(self.sequence, other.sequence, l, u)
			return Sequence(sq.toString(res[0], res[1], res[2]))

	def __imul__(self, other): #f(n) *= f(n)
		l = min(self.lower, other.lower)
		u = max(self.upper, other.upper)
		res = sq.subSeqs(self.sequence, other.sequence, l, u)
		return Sequence(sq.toString(res[0], res[1], res[2]))

	def __neg__(self): #f(-n)
		res = sq.negSeq(self.sequence, self.lower, self.upper)
		return Sequence(sq.toString(res, list(res)[0], list(res)[-1]))

File: resources/sequence.js
import * as sq from './seq';

class Sequence {
	constructor(string) {
		this.sequence = sq.createSeq(string);
		this.lower =  parseInt(Object.keys(this.sequence).sort((a, b) => {
			return parseInt(a) - parseInt(b);
		})[0]);
		this.upper = parseInt(Object.keys(this.sequence).sort((a, b) => {
			return parseInt(a) - parseInt(b);
		}).pop()); //Nuevo posicion
	}

	get getSequence(){
		return this.sequence
	}

	get getString(){
		return sq.toString(this.sequence, this.lower, this.upper)
	}

	get conSequence(g){ //Convolucion
		res = sq.conSeq(this.sequence, this.lower, this.upper, g, g.lower, g.upper)
		let lower =  parseInt(Object.keys(res).sort((a, b) => {
			return parseInt(a) - parseInt(b);
		})[0]);
		let upper = parseInt(Object.keys(res).sort((a, b) => {
			return parseInt(a) - parseInt(b);
		}).pop()); //Nuevo posicion
		return new Sequence(sq.toString(res, lower, upper))
	}

	add(other){ //#f(n) += f(n)
		if (typeof other === 'number'){ //Cuando es desplazamiento
			res = sq.desSeq(this.sequence, this.lower, this.upper, other)
			return Sequence(sq.toString(res, list(res)[0], list(res)[-1]))
		}
		l = Math.min(this.lower, other.lower)
		u = Math.max(this.upper, other.upper)
		res = sq.addSeqs(this.sequence, other.sequence, l, u)
		return new Sequence(sq.toString(res[0], res[1], res[2]))
	}

	static add(a, b){ //#f(n) += f(n)
		let res = null;
		if (typeof a === 'number'){ //Cuando es desplazamiento
			res = sq.desSeq(b.sequence, b.lower, b.upper, a);
		}else if (typeof b === 'number') //Cuando es desplazamiento
			res = sq.desSeq(a.sequence, a.lower, a.upper, b);
		
		if (res) return Sequence(sq.toString(res, parseInt(Object.keys(res).sort((a, b) => {
				return parseInt(a) - parseInt(b);
			})[0]),  parseInt(Object.keys(res).sort((a, b) => {
				return parseInt(a) - parseInt(b);
			}).pop())));

		let l = Math.min(a.lower, b.lower)
		let u = Math.max(a.upper, b.upper)
		res = sq.addSeqs(a.sequence, b.sequence, l, u)
		res = new Sequence(sq.toString(res[0], res[1], res[2]))
	}
}